tests fail on Go 1.5.1

Hey there, I was looking into your bcast library and noticed some tests fail on my machine with Go 1.5.1:

```
dq@empit:bcast$ go test -v
=== RUN   TestNewGroupAndJoin
--- PASS: TestNewGroupAndJoin (0.00s)
=== RUN   TestUnjoin
--- PASS: TestUnjoin (0.00s)
=== RUN   TestBroadcast
--- FAIL: TestBroadcast (0.10s)
    bcast_test.go:85: not all messages broadcasted
=== RUN   TestBroadcastFromMember
--- FAIL: TestBroadcastFromMember (0.10s)
    bcast_test.go:114: not all messages broadcasted
=== RUN   TestGroupBroadcast
--- FAIL: TestGroupBroadcast (0.15s)
    bcast_test.go:141: not all messages broadcasted
=== RUN   TestBroadcastOnLargeNumberOfMembers
--- FAIL: TestBroadcastOnLargeNumberOfMembers (1.05s)
    bcast_test.go:168: not all messages broadcasted (11267/16256)
FAIL
exit status 1
FAIL    github.com/grafov/bcast 1.415s
dq@empit:bcast$ go version
go version go1.5.1 linux/amd64
```

cheers/

--- a/bcast.go
+++ b/bcast.go
@@ -1,4 +1,3 @@
-// Broadcast over channels.
 package bcast
 
 /*
@@ -11,180 +10,206 @@ package bcast
 */
 
 import (
+	"container/heap"
+	"errors"
 	"sync"
 	"time"
-	//"fmt"
 )
 
-// Internal structure to pack messages together with info about sender.
+// Message is an internal structure to pack messages together with
+// info about sender.
 type Message struct {
-	sender  chan interface{}
+	sender  *Member
 	payload interface{}
+	clock   int
 }
 
-// Represents member of broadcast group.
+// Member represents member of a Broadcast group.
 type Member struct {
-	group *Group           // send messages to others directly to group.In
-	In    chan interface{} // (public) get messages from others to own channel
-	//NonReceived int
+	group        *Group
+	Read         chan interface{}
+	clock        int
+	messageQueue PriorityQueue
+	send         chan Message
+	close        chan bool
 }
 
-// Represents broadcast group.
+// Group provides a mechanism for the broadcast of messages to a
+// collection of channels.
 type Group struct {
-	l     sync.Mutex
-	in    chan Message       // receive broadcasts from members
-	out   []chan interface{} // broadcast messages to members
-	count int
-	close chan bool
+	in         chan Message
+	close      chan bool
+	members    []*Member
+	clock      int
+	memberLock sync.Mutex
+	clockLock  sync.Mutex
 }
 
-// Create new broadcast group.
+// NewGroup creates a new broadcast group.
 func NewGroup() *Group {
 	in := make(chan Message)
 	close := make(chan bool)
-	return &Group{in: in, close: close}
+	return &Group{in: in, close: close, clock: 0}
 }
 
-func (r *Group) MemberCount() int {
-	return r.count
+// MemberCount returns the number of members in the Broadcast Group.
+func (g *Group) MemberCount() int {
+	return len(g.Members())
 }
 
-func (r *Group) Members() []chan interface{} {
-	r.l.Lock()
-	res := r.out[:]
-	r.count = len(r.out)
-	r.l.Unlock()
+// Members returns a slice of Members that are currently in the Group.
+func (g *Group) Members() []*Member {
+	g.memberLock.Lock()
+	res := g.members[:]
+	g.memberLock.Unlock()
 	return res
 }
 
-func (r *Group) Add(out chan interface{}) {
-	r.l.Lock()
-	r.out = append(r.out, out)
-	r.count = len(r.out)
-	r.l.Unlock()
-	return
+// Join returns a new member object and handles the creation of its
+// output channel.
+func (g *Group) Join() *Member {
+	memberChannel := make(chan interface{})
+	return g.Add(memberChannel)
 }
 
-func (r *Group) Remove(received Message) {
-	r.l.Lock()
-	for i, addr := range r.out {
-		if addr == received.payload.(Member).In && received.sender == received.payload.(Member).In {
-			r.out = append(r.out[:i], r.out[i+1:]...)
-			r.count = len(r.out)
+// Leave removes the provided member from the group
+func (g *Group) Leave(leaving *Member) error {
+	g.memberLock.Lock()
+	memberIndex := -1
+	for index, member := range g.members {
+		if member == leaving {
+			memberIndex = index
 			break
 		}
 	}
-	r.l.Unlock()
-	return
+	if memberIndex == -1 {
+		return errors.New("Could not find provided memeber for removal")
+	}
+	g.members = append(g.members[:memberIndex], g.members[memberIndex+1:]...)
+	leaving.close <- true // TODO: need to handle the case where there
+	// is still stuff in this Members priorityQueue
+	g.memberLock.Unlock()
+	return nil
 }
 
-// Close the group immediately
-func (r *Group) Close() {
-	r.close <- true
+// Add adds a member to the group for the provided interface channel.
+func (g *Group) Add(memberChannel chan interface{}) *Member {
+	g.memberLock.Lock()
+	g.clockLock.Lock()
+	member := &Member{
+		group:        g,
+		Read:         memberChannel,
+		clock:        g.clock,
+		messageQueue: PriorityQueue{},
+		send:         make(chan Message),
+		close:        make(chan bool),
+	}
+	go member.listen()
+	g.members = append(g.members, member)
+	g.clockLock.Unlock()
+	g.memberLock.Unlock()
+	return member
 }
 
-// Obsoleted by BroadcastFor()
-func (r *Group) Broadcasting(timeout time.Duration) {
-	r.BroadcastFor(timeout)
+// Close terminates the group immediately.
+func (g *Group) Close() {
+	g.close <- true
 }
 
 // Broadcast messages received from one group member to others.
 // If incoming messages not arrived during `timeout` then function returns.
-func (r *Group) BroadcastFor(timeout time.Duration) {
-	if timeout == 0 {
-		r.Broadcast()
-		return
+func (g *Group) Broadcast(timeout time.Duration) {
+	var timeoutChannel <-chan time.Time
+	if timeout != 0 {
+		timeoutChannel = time.After(timeout)
 	}
 	for {
 		select {
-		case received := <-r.in:
-			switch received.payload.(type) {
-			default: // receive a payload and broadcast it
-				for _, member := range r.Members() {
-					if received.sender != member { // not return broadcast to sender
-
-						/*
-							select {
-							case member <- received.payload:
-								fmt.Println("sent message", received.payload)
-							default:
-								fmt.Println("no message sent")
-							}
-						*/
-						go func(out chan interface{}, received *Message) { // non blocking
-							out <- received.payload
-						}(member, &received)
-
-					}
-				}
+		case received := <-g.in:
+			g.memberLock.Lock()
+			g.clockLock.Lock()
+			members := g.members[:]
+			received.clock = g.clock
+			g.clock++
+			g.clockLock.Unlock()
+			g.memberLock.Unlock()
+			for _, member := range members {
+				// This is done in a goroutine because if it
+				// weren't it would be a blocking call
+				go func(member *Member, received Message) {
+					member.send <- received
+				}(member, received)
 			}
-		case <-time.After(timeout):
+		case <-timeoutChannel:
 			if timeout > 0 {
 				return
 			}
-		case <-r.close:
+		case <-g.close:
 			return
 		}
 	}
 }
 
-// Broadcast messages received from one group member to others.
-// See https://github.com/grafov/bcast/issues/4 for rationale.
-func (r *Group) Broadcast() {
+// Send broadcasts a message to every one of a Group's members.
+func (g *Group) Send(val interface{}) {
+	g.in <- Message{sender: nil, payload: val}
+}
+
+// Close removes the member it is called on from its broadcast group.
+func (m *Member) Close() {
+	m.group.Leave(m)
+}
+
+// Send broadcasts a message from one Member to the channels of all
+// the other members in its group.
+func (m *Member) Send(val interface{}) {
+	m.group.in <- Message{sender: m, payload: val}
+}
+
+// Recv reads one value from the member's Read channel
+func (m *Member) Recv() interface{} {
+	return <-m.Read
+}
+
+func (m *Member) listen() {
 	for {
 		select {
-		case <-r.close:
+		case message := <-m.send:
+			m.handleMessage(&message)
+		case <-m.close:
 			return
-		case received := <-r.in:
-			switch received.payload.(type) {
-			default: // receive a payload and broadcast it
-				for _, member := range r.Members() {
-					if received.sender != member { // not return broadcast to sender
+		}
+	}
+}
 
-						/*
-							select {
-							case member <- received.payload:
-								//fmt.Println("sent message", received.payload)
-							default:
-								//fmt.Println("no message sent")
-							}
-						*/
-						go func(out chan interface{}, received *Message) { // non blocking
-							out <- received.payload
-						}(member, &received)
-					}
-				}
+func (m *Member) handleMessage(message *Message) {
+	if !m.trySend(message) {
+		heap.Push(&m.messageQueue, &Item{
+			priority: message.clock,
+			value:    message,
+		})
+		return
+	}
+	if m.messageQueue.Len() > 0 {
+		nextMessage := m.messageQueue[0].value.(*Message)
+		for m.trySend(nextMessage) {
+			heap.Pop(&m.messageQueue)
+			if m.messageQueue.Len() > 0 {
+				nextMessage = m.messageQueue[0].value.(*Message)
+			} else {
+				break
 			}
 		}
 	}
 }
 
-// Broadcast message to all group members.
-func (r *Group) Send(val interface{}) {
-	r.in <- Message{sender: nil, payload: val}
-}
-
-// Join new member to broadcast.
-func (r *Group) Join() *Member {
-	out := make(chan interface{})
-	r.Add(out)
-	//r.out = append(r.out, out)
-	return &Member{group: r, In: out}
-}
-
-// Unjoin member from broadcast group.
-func (r *Member) Close() {
-	r.group.Remove(Message{sender: r.In, payload: *r})
-	//r.group.in <- Message{sender: r.In, payload: *r} // broadcasting of self means member closing
-}
-
-// Broadcast message from one member to others except sender.
-func (r *Member) Send(val interface{}) {
-	r.group.in <- Message{sender: r.In, payload: val}
-}
-
-// Get broadcast message.
-// As alternative you may get it from `In` channel.
-func (r *Member) Recv() interface{} {
-	return <-r.In
+func (m *Member) trySend(message *Message) bool {
+	shouldSend := message.clock == m.clock
+	if shouldSend {
+		if message.sender != m {
+			m.Read <- message.payload
+		}
+		m.clock++
+	}
+	return shouldSend
 }

--- a/bcast_test.go
+++ b/bcast_test.go
@@ -10,9 +10,9 @@ package bcast
 */
 
 import (
+	"gopkg.in/fatih/set.v0"
 	"testing"
 	"time"
-	"sync"
 )
 
 // Create new broadcast group.
@@ -33,85 +33,58 @@ func TestUnjoin(t *testing.T) {
 	group := NewGroup()
 	member1 := group.Join()
 	member2 := group.Join()
-	if len(group.out) != 2 {
+	if len(group.members) != 2 {
 		t.Fatal("incorrect length of `out` slice")
 	}
-
-	go group.Broadcasting(2 * time.Second)
+	go group.Broadcast(2 * time.Second)
 
 	member1.Close()
-	if len(group.out) > 1 || group.out[0] != member2.In {
+	if len(group.members) > 1 || group.members[0] != member2 {
 		t.Fatal("unjoin member does not work")
 	}
 }
 
-type Adder struct {
-	count int
-	l sync.Mutex
-}
-
-func (a *Adder) Add(c int) {
-	a.l.Lock()
-	a.count += c
-	a.l.Unlock()
-}
-
 // Create new broadcast group.
 // Join 12 members.
-// Broadcast one integer from each member.
-func TestBroadcast(t *testing.T) {
-	var valcount Adder
-	valcount.count = 0
-
+// Make group broadcast to all group members.
+func TestMemberBroadcast(t *testing.T) {
 	group := NewGroup()
+	var channels []chan bool
+	max := 12
+	broadcaster := 3
 
-	for i := 1; i <= 12; i++ {
-		go func(i int) {
-			m := group.Join()
-			m.Send(i)
+	for i := 0; i <= max; i++ {
+		c := make(chan bool)
+		channels = append(channels, c)
+	}
 
-			for {
+	for i, c := range channels {
+		m := group.Join()
+		go func(i int, group *Group, channel chan bool, member *Member) {
+			if i == broadcaster {
+				m.Send(i)
+			} else {
 				val := m.Recv()
-				if val.(int) == i {
-					t.Fatal("sent value was received by sender")
+				if val != broadcaster {
+					t.Fatal("incorrect message received")
 				}
-				valcount.Add(1)
 			}
-		}(i)
-	}
-
-	group.Broadcasting(100 * time.Millisecond)
-	if valcount.count != 12*12-12 { // number of channels * number of messages - number of channels
-		t.Fatal("not all messages broadcasted")
-	}
-}
-
-// Create new broadcast group.
-// Join 12 members.
-// Broadcast one integer from each member.
-func TestBroadcastFromMember(t *testing.T) {
-	var valcount int
-
-	group := NewGroup()
-
-	for i := 1; i <= 12; i++ {
-		go func(i int, group *Group) {
-			m := group.Join()
-			m.Send(i)
-
-			for {
-				val := m.Recv()
-				if val.(int) == i {
-					t.Fatal("sent value was received by sender")
-				}
-				valcount++
+			channel <- true
+			val := m.Recv()
+			if val != "done" {
+				t.Fatal("incorrect message received")
 			}
-		}(i, group)
+			channel <- true
+		}(i, group, c, m)
 	}
 
-	group.Broadcasting(100 * time.Millisecond)
-	if valcount != 12*12-12 { // number of channels * number of messages - number of channels
-		t.Fatal("not all messages broadcasted")
+	go group.Broadcast(0)
+	for _, channel := range channels {
+		<-channel
+	}
+	group.Send("done")
+	for _, channel := range channels {
+		<-channel
 	}
 }
 
@@ -119,26 +92,30 @@ func TestBroadcastFromMember(t *testing.T) {
 // Join 12 members.
 // Make group broadcast to all group members.
 func TestGroupBroadcast(t *testing.T) {
-	var valcount int
-
 	group := NewGroup()
+	var channels []chan bool
+	max := 12
 
-	for i := 1; i <= 12; i++ {
-		go func(i int, group *Group) {
-			m := group.Join()
-			if val := m.Recv(); val != "group message" {
-				t.Fatal("incorrect message received")
-			}
-			valcount++
-		}(i, group)
+	for i := 0; i <= max; i++ {
+		c := make(chan bool)
+		channels = append(channels, c)
 	}
 
-	go group.Broadcasting(200 * time.Millisecond)
-	group.Send("group message")
-	time.Sleep(150 * time.Millisecond)
+	for i, c := range channels {
+		m := group.Join()
+		go func(i int, group *Group, channel chan bool, member *Member) {
+			val := m.Recv()
+			if val != "group message" {
+				t.Fatal("incorrect message received")
+			}
+			channel <- true
+		}(i, group, c, m)
+	}
 
-	if valcount != 12 {
-		t.Fatal("not all messages broadcasted")
+	go group.Broadcast(0)
+	group.Send("group message")
+	for _, channel := range channels {
+		<-channel
 	}
 }
 
@@ -147,24 +124,38 @@ func TestGroupBroadcast(t *testing.T) {
 // Make group broadcast to all members.
 func TestBroadcastOnLargeNumberOfMembers(t *testing.T) {
 	const max = 128
-	var valcount int
+	var channels []chan *set.Set
+	var members []*Member
+	expected := set.New()
 
 	group := NewGroup()
-	for i := 1; i <= max; i++ {
-		go func(i int, group *Group) {
-			m := group.Join() // here is the problem
-			m.Send(i)
-			for {
-				if val := m.Recv(); val.(int) == i {
-					t.Fatal("sent value should not received by sender")
-				}
-				valcount++
-			}
-		}(i, group)
+	for i := 0; i <= max; i++ {
+		expected.Add(i)
+		m := group.Join()
+		members = append(members, m)
 	}
-
-	group.Broadcasting(1 * time.Second)
-	if valcount != max*max-max { // number of channels * number of messages - number of channels
-		t.Fatalf("not all messages broadcasted (%d/%d)", valcount, max*max-max)
+	for i, member := range members {
+		c := make(chan *set.Set)
+		channels = append(channels, c)
+		go func(i int, group *Group, channel chan *set.Set, m *Member) {
+			m.Send(i)
+			encountered := set.New()
+			encountered.Add(i) // The message sent by this member wont be received
+			for {
+				newValue := m.Recv()
+				if encountered.Has(newValue) {
+					t.Fatal("Received duplicate value")
+				}
+				encountered.Add(newValue)
+				if encountered.IsEqual(expected) {
+					break
+				}
+			}
+			channel <- encountered
+		}(i, group, c, member)
+	}
+	go group.Broadcast(0)
+	for _, channel := range channels {
+		<-channel
 	}
 }

--- a/priority_queue.go
+++ b/priority_queue.go
@@ -1,0 +1,52 @@
+package bcast
+
+import (
+	"container/heap"
+)
+
+// An Item is something we manage in a priority queue.
+type Item struct {
+	value    interface{}
+	priority int // The priority of the item in the queue.
+	// The index is needed by update and is maintained by the heap.Interface methods.
+	index int // The index of the item in the heap.
+}
+
+// A PriorityQueue implements heap.Interface and holds Items.
+type PriorityQueue []*Item
+
+func (pq PriorityQueue) Len() int { return len(pq) }
+
+func (pq PriorityQueue) Less(i, j int) bool {
+	// We want Pop to give us the lowest priority so we use less than here.
+	return pq[i].priority < pq[j].priority
+}
+
+func (pq PriorityQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *PriorityQueue) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*Item)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+func (pq *PriorityQueue) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+// update modifies the priority and value of an Item in the queue.
+func (pq *PriorityQueue) update(item *Item, value string, priority int) {
+	item.value = value
+	item.priority = priority
+	heap.Fix(pq, item.index)
+}

